release-25.4: sql: log when optimizer estimates for scans are inaccurate

Uzair5162 opened the pull request:
Backport 3/3 commits from #154370 on behalf of @Uzair5162.
Backport 1/1 commits from #155454 on behalf of @Uzair5162.

----

#### sql: estimate table statistic staleness in stats.Refresher
This commit adds an `EstimateStaleness()` method to the table statistic
`Refresher`, which estimates the current fraction of stale rows in a
given table with the formula:

`cur_fraction_stale = (time_since_last_refresh /
avg_time_between_refreshes) * target_fraction_stale_rows`

Although this isn’t used anywhere yet, it will be useful for logging
when scans are misestimated (see https://github.com/cockroachdb/cockroach/issues/153748).

Part of: https://github.com/cockroachdb/cockroach/issues/153748, https://github.com/cockroachdb/cockroach/issues/153873

Release note: None

#### sql: plumb StageID into table reader processors

This commit plumbs `StageID` from the general `ProcessorSpec` into
table reader processors. Specifically, `StageID` is plumbed into
`tableReader` for the row-based flow and `colBatchScanBase` (via
`ColBatchScan` and `ColBatchDirectScan`) for the vectorized flow.
Although `stageID` isn't used in these processors yet, it will
be useful for aggregating row counts from metrics across distributed
table readers for misestimate logging.

Part of: https://github.com/cockroachdb/cockroach/issues/153748

Release note: None

#### sql: log when optimizer estimates for scans are inaccurate
This commit logs a warning on the gateway node when the estimated row
count for a logical scan is inaccurate. The `DistSQLReceiver` on the
gateway node now maintains the row count estimate and metadata for each
logical scan stage. Table reader metrics are extended to include their
StageID, which we use to aggregate the emitted row counts from all table
readers processors corresponding to the same plan stage at the receiver.

An estimate is considered inaccurate if it is off by at least a factor
of 2 and a fixed offset of 100, matching the logic in the warning from
`EXPLAIN ANALYZE`. The log message includes the table and index being
scanned, the estimated and actual row counts, the time since the last
table stats collection, and the table's estimated staleness.

This log is gated behind a new cluster setting,
`sql.log.scan_row_count_misestimate.enabled` (default off). Logging only
happens for user tables and is rate limited to log misestimates from at
most 1 query every 10 seconds.

Fixes: https://github.com/cockroachdb/cockroach/issues/153748
Fixes: https://github.com/cockroachdb/cockroach/issues/153873

Release note (sql change): Added a default-off cluster setting
(`sql.log.scan_row_count_misestimate.enabled`) that enables logging a
warning on the gateway node when optimizer estimates for scans are
inaccurate. The log message includes the table and index being scanned,
the estimated and actual row counts, the time since the last table stats
collection, and the table's estimated staleness.

#### sql: optimizer scan misestimate logging improvements
This commit is a follow-up to https://github.com/cockroachdb/cockroach/pull/154370. It changes misestimate logging to
use structured logging on the `SQL_EXEC` log channel with a new
`ScanRowCountMisestimate` event, instead of unstructured logs on DEV. It
also adds an error case to `stats.EstimateStaleness()` when there are 1
or fewer full automatic stats collections, and uses the fully qualified
table name in logs.

Part of: https://github.com/cockroachdb/cockroach/issues/153748, https://github.com/cockroachdb/cockroach/issues/153873

Release note (sql change): Changes scan misestimate logging gated behind
`sql.log.scan_row_count_misestimate.enabled` to use structured logging
including the table and index being scanned, the estimated and actual
row counts, the time since the last table stats collection, and the
table's estimated staleness.

----

Release justification: resolve GA-blocker, gated behind disabled-by-default `sql.log.scan_row_count_misestimate.enabled`

Uzair5162 committed:
sql: estimate table statistic staleness in `stats.Refresher`

This commit adds an `EstimateStaleness()` method to the table statistic
`Refresher`, which estimates the current fraction of stale rows in a
given table with the formula:

`cur_fraction_stale = (time_since_last_refresh /
avg_time_between_refreshes) * target_fraction_stale_rows`

Although this isn’t used anywhere yet, it will be useful for logging
when scans are misestimated (see #153748).

Part of: #153748, #153873

Release note: None

Uzair5162 committed:
sql: plumb StageID into table reader processors

This commit plumbs `StageID` from the general `ProcessorSpec` into
table reader processors. Specifically, `StageID` is plumbed into
`tableReader` for the row-based flow and `colBatchScanBase` (via
`ColBatchScan` and `ColBatchDirectScan`) for the vectorized flow.
Although `stageID` isn't used in these processors yet, it will
be useful for aggregating row counts from metrics across distributed
table readers for misestimate logging.

Part of: #153748

Release note: None

Uzair5162 committed:
sql: log when optimizer estimates for scans are inaccurate

This commit logs a warning on the gateway node when the estimated row
count for a logical scan is inaccurate. The `DistSQLReceiver` on the
gateway node now maintains the row count estimate and metadata for each
logical scan stage. Table reader metrics are extended to include their
StageID, which we use to aggregate the emitted row counts from all table
readers processors corresponding to the same plan stage at the receiver.

An estimate is considered inaccurate if it is off by at least a factor
of 2 and a fixed offset of 100, matching the logic in the warning from
`EXPLAIN ANALYZE`. The log message includes the table and index being
scanned, the estimated and actual row counts, the time since the last
table stats collection, and the table's estimated staleness.

This log is gated behind a new cluster setting,
`sql.log.scan_row_count_misestimate.enabled` (default off). Logging only
happens for user tables and is rate limited to log misestimates from at
most 1 query every 10 seconds.

Fixes: #153748, #153873

Release note (sql change): Added a default-off cluster setting
(`sql.log.scan_row_count_misestimate.enabled`) that enables logging a
warning on the gateway node when optimizer estimates for scans are
inaccurate. The log message includes the table and index being scanned,
the estimated and actual row counts, the time since the last table stats
collection, and the table's estimated staleness.

michae2 reviewed (commented):
Nice work! I have some comments, but we can discuss them a bit today since we're close to eng completion.

> @michae2 reviewed 2 of 2 files at r1, 13 of 13 files at r2, 17 of 18 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/155123#-:-ObYafpvEvj4hlq8YaQ3:b-g4dcm8)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @rytaft)

___
*[`pkg/sql/stats/automatic_stats.go` line 938 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/155123#-ObF-Umt4vFTxLcNJZzf:-ObF-Umt4vFTxLcNJZzg:bwkei90) ([raw file](https://github.com/cockroachdb/cockroach/blob/04441b146a0b5ea6c8f322e9f9dba724a9c54564/pkg/sql/stats/automatic_stats.go#L938)):*
> ```Go
> 	staleTargetFraction := r.autoStatsFractionStaleRows(explicitSettings)
> 
> 	avgRefreshTime := avgFullRefreshTime(tableStats)
> ```

avgFullRefreshTime will give a default time if there are 1 or fewer full automatic stats collections. That might produce kind of a strange staleFraction. It might be worth returning an error if there are 1 or fewer full automatic stats collections to avoid that case.
___
*[`pkg/sql/distsql_running.go` line 1361 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/155123#-ObFG0j0CJQamASKt1mm:-ObFG0j0CJQamASKt1mn:b-mvf8i3) ([raw file](https://github.com/cockroachdb/cockroach/blob/7fdc1b1022a35345dc3f6660d791b7b788141af8/pkg/sql/distsql_running.go#L1361)):*
> ```Go
> 	statsCreatedAt    time.Time
> 	tableID           descpb.ID
> 	tableName         string
> ```

Is this the fully-qualified table name? Or just the last part? We should be logging the fully-qualified table name.

(It might be interesting to log the misestimated span(s), too, but that's less important.)
___
*[`pkg/sql/distsql_running.go` line 1367 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/155123#-ObSqFTe16aPxJ8uBO68:-ObSqFTe16aPxJ8uBO69:bynfhx2) ([raw file](https://github.com/cockroachdb/cockroach/blob/7fdc1b1022a35345dc3f6660d791b7b788141af8/pkg/sql/distsql_running.go#L1367)):*
> ```Go
> }
> 
> var misestimateLogLimiter = log.Every(10 * time.Second)
> ```

I'm glad we have this, but the fact that it's global rather than per-table means there is a chance that a single noisy table could accidentally drown out log messages from other tables with misestimates. I wonder if there's a way to make the limiter per-table rather than global?
___
*[`pkg/sql/distsql_running.go` line 1380 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/155123#-ObFGrmSEGyjvGzM9S3i:-ObFGrmSEGyjvGzM9S3j:bsr7x4i) ([raw file](https://github.com/cockroachdb/cockroach/blob/7fdc1b1022a35345dc3f6660d791b7b788141af8/pkg/sql/distsql_running.go#L1380)):*
> ```Go
> 		}
> 	}
> 	log.Dev.Warningf(ctx, "inaccurate estimate for scan on table %q (index %q): estimated=%d actual=%d%s",
> ```

Ideally, for logs we expect customers to build alerting on, we would be using structured logging in a different log channel than DEV (maybe SQL_EXEC, for example see SlowQuery in pkg/util/log/eventpb/sql_audit_events.proto). It's hard for customers to consume unstructured logging in DEV because (a) they have to parse the log message and (b) the DEV channel is extremely noisy and not intended to be customer-facing.
___
*[`pkg/sql/distsql_running.go` line 1494 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/155123#-ObFIDIr33sMwCf8Qw79:-ObFIDIr33sMwCf8Qw7A:bcwlmnp) ([raw file](https://github.com/cockroachdb/cockroach/blob/7fdc1b1022a35345dc3f6660d791b7b788141af8/pkg/sql/distsql_running.go#L1494)):*
> ```Go
> 		stmtType:             tree.Rows,
> 		tracing:              r.tracing,
> 		scanStageEstimateMap: ret.scanStageEstimateMap,
> ```

Should we be referencing the map like this? (In other words, do sub- and post- queries have unique stage IDs within the root DistSQL plan, or do they have their own stage IDs independent of the root plan?)
___
*[`pkg/sql/distsql_running.go` line 1864 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/155123#-ObFMAgW8qacROS8pnXM:-ObFMAgW8qacROS8pnXN:bjao8n3) ([raw file](https://github.com/cockroachdb/cockroach/blob/7fdc1b1022a35345dc3f6660d791b7b788141af8/pkg/sql/distsql_running.go#L1864)):*
> ```Go
> 		estimatedRowCount := s.estimatedRowCount
> 		if estimatedRowCount == 0 {
> 			continue
> ```

Why do we not log if the estimatedRowCount is zero?


<!-- Sent from Reviewable.io -->

yuzefovich reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/155123)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @michae2, @rytaft, and @Uzair5162)

___
*[`pkg/sql/distsql_running.go` line 1494 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/155123#-ObFIDIr33sMwCf8Qw79:-ObYkZytBaQejHOHOHOq:b-lteawt) ([raw file](https://github.com/cockroachdb/cockroach/blob/7fdc1b1022a35345dc3f6660d791b7b788141af8/pkg/sql/distsql_running.go#L1494)):*
<details><summary><i>Previously, michae2 (Michael Erickson) wrote…</i></summary><blockquote>

Should we be referencing the map like this? (In other words, do sub- and post- queries have unique stage IDs within the root DistSQL plan, or do they have their own stage IDs independent of the root plan?)
</blockquote></details>

The map here is empty and it's reused when we're getting the object from `receiverSyncPool`. (Sub- and post-queries don't have unique stage IDs, but we don't inherit the map from the receiver of the main query, note we use `ret` not `r` here.)
___
*[`pkg/sql/distsql_running.go` line 1864 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/155123#-ObFMAgW8qacROS8pnXM:-ObYkusH6RQJBndCAMe9:b4zhoay) ([raw file](https://github.com/cockroachdb/cockroach/blob/7fdc1b1022a35345dc3f6660d791b7b788141af8/pkg/sql/distsql_running.go#L1864)):*
<details><summary><i>Previously, michae2 (Michael Erickson) wrote…</i></summary><blockquote>

Why do we not log if the estimatedRowCount is zero?
</blockquote></details>

The `estimatedRowCount` is zero when the table doesn't have stats (as well as when the actual estimate is zero). We'd need to change how we propagate this estimate (which will require some changes elsewhere too) if we want to distinguish these two cases - do you think the actual estimate of zero is an important case to not ignore?


<!-- Sent from Reviewable.io -->

michae2 reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/155123)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @rytaft and @Uzair5162)

___
*[`pkg/sql/distsql_running.go` line 1494 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/155123#-ObFIDIr33sMwCf8Qw79:-ObZLe4uEXyVjVnZdxJR:b-89g8s1) ([raw file](https://github.com/cockroachdb/cockroach/blob/7fdc1b1022a35345dc3f6660d791b7b788141af8/pkg/sql/distsql_running.go#L1494)):*
<details><summary><i>Previously, yuzefovich (Yahor Yuzefovich) wrote…</i></summary><blockquote>

The map here is empty and it's reused when we're getting the object from `receiverSyncPool`. (Sub- and post-queries don't have unique stage IDs, but we don't inherit the map from the receiver of the main query, note we use `ret` not `r` here.)
</blockquote></details>

Ah, I missed that it was `ret` and not `r`! Sorry! Thanks for explaining!
___
*[`pkg/sql/distsql_running.go` line 1864 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/155123#-ObFMAgW8qacROS8pnXM:-ObZLssE7Dy_Y_jufFf3:b-vj7qrk) ([raw file](https://github.com/cockroachdb/cockroach/blob/7fdc1b1022a35345dc3f6660d791b7b788141af8/pkg/sql/distsql_running.go#L1864)):*
<details><summary><i>Previously, yuzefovich (Yahor Yuzefovich) wrote…</i></summary><blockquote>

The `estimatedRowCount` is zero when the table doesn't have stats (as well as when the actual estimate is zero). We'd need to change how we propagate this estimate (which will require some changes elsewhere too) if we want to distinguish these two cases - do you think the actual estimate of zero is an important case to not ignore?
</blockquote></details>

Ah, I didn't realize we always round estimated row counts up, so this should always be >= 1 if the stats are available and epsilon was used in statistics builder. This is probably fine, then.


<!-- Sent from Reviewable.io -->

Uzair5162 committed:
sql: optimizer scan misestimate logging improvements

This commit is a follow-up to #154370. It changes misestimate logging to
use structured logging on the `SQL_EXEC` log channel with a new
`ScanRowCountMisestimate` event, instead of unstructured logs on DEV. It
also adds an error case to `stats.EstimateStaleness()` when there are 1
or fewer full automatic stats collections, and uses the fully qualified
table name in logs.

Part of: #153748, #153873

Release note (sql change): Changes scan misestimate logging gated behind
`sql.log.scan_row_count_misestimate.enabled` to use structured logging
including the table and index being scanned, the estimated and actual
row counts, the time since the last table stats collection, and the
table's estimated staleness.

Uzair5162 reviewed (commented):
TFTR! I've made the suggested changes on https://github.com/cockroachdb/cockroach/pull/155454, except for per-table logging since I think that might be a bigger change that can be explored post-release.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/155123#-:-ObcjSgh9xcANwBdP1ma:b-knae6h)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @rytaft)

___
*[`pkg/sql/distsql_running.go` line 1361 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/155123#-ObFG0j0CJQamASKt1mm:-ObckSzK6klUTxBTOLGR:b-ubix0w) ([raw file](https://github.com/cockroachdb/cockroach/blob/7fdc1b1022a35345dc3f6660d791b7b788141af8/pkg/sql/distsql_running.go#L1361)):*
<details><summary><i>Previously, michae2 (Michael Erickson) wrote…</i></summary><blockquote>

Is this the fully-qualified table name? Or just the last part? We should be logging the fully-qualified table name.

(It might be interesting to log the misestimated span(s), too, but that's less important.)
</blockquote></details>

This is not currently the fully-qualified table name. I've changed it to use the fully qualified name in https://github.com/cockroachdb/cockroach/pull/155454.
___
*[`pkg/sql/distsql_running.go` line 1380 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/155123#-ObFGrmSEGyjvGzM9S3i:-ObckgZC6a6UKc7T33um:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/7fdc1b1022a35345dc3f6660d791b7b788141af8/pkg/sql/distsql_running.go#L1380)):*
<details><summary><i>Previously, michae2 (Michael Erickson) wrote…</i></summary><blockquote>

Ideally, for logs we expect customers to build alerting on, we would be using structured logging in a different log channel than DEV (maybe SQL_EXEC, for example see SlowQuery in pkg/util/log/eventpb/sql_audit_events.proto). It's hard for customers to consume unstructured logging in DEV because (a) they have to parse the log message and (b) the DEV channel is extremely noisy and not intended to be customer-facing.
</blockquote></details>

Done.
___
*[`pkg/sql/stats/automatic_stats.go` line 938 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/155123#-ObF-Umt4vFTxLcNJZzf:-Obc4R9UEzN2Za2bt0UR:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/04441b146a0b5ea6c8f322e9f9dba724a9c54564/pkg/sql/stats/automatic_stats.go#L938)):*
<details><summary><i>Previously, michae2 (Michael Erickson) wrote…</i></summary><blockquote>

avgFullRefreshTime will give a default time if there are 1 or fewer full automatic stats collections. That might produce kind of a strange staleFraction. It might be worth returning an error if there are 1 or fewer full automatic stats collections to avoid that case.
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

Uzair5162 reviewed (commented):
I've updated this to include the commit from https://github.com/cockroachdb/cockroach/pull/155454.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/155123#-:-ObhBPGy-vRM58BTKvTO:bqxajeg)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @michae2 and @rytaft)



<!-- Sent from Reviewable.io -->

michae2 reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> Great work!

> @michae2 reviewed 7 of 7 files at r4, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/155123#-:-ObhyVmE7zXIfCnGHVdD:bolbkdq)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (waiting on @rytaft)



<!-- Sent from Reviewable.io -->
